File: backend/app/schemas/curriculum.py
import uuid

from pydantic import BaseModel, ConfigDict


class SkillRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    name: str
    slug: str
    category: str | None = None


class TopicRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    name: str
    slug: str


class LessonSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    title: str
    slug: str
    order: int
    estimated_minutes: int | None = None


class LessonDetail(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    title: str
    slug: str
    order: int
    content: dict
    estimated_minutes: int | None = None
    skills: list[SkillRead] = []
    quiz_id: uuid.UUID | None = None


class ModuleWithLessons(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    title: str
    slug: str
    order: int
    lessons: list[LessonSummary] = []


class CourseSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    title: str
    slug: str
    description: str | None = None
    level: str
    estimated_hours: int | None = None


class CourseDetail(CourseSummary):
    modules: list[ModuleWithLessons] = []
    completed_lesson_ids: list[uuid.UUID] = []


class LearningPathSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    title: str
    slug: str
    description: str | None = None


class LearningPathDetail(LearningPathSummary):
    courses: list[CourseSummary] = []


class QuizQuestionPublic(BaseModel):
    """Correct answers are deliberately excluded — the client must never
    receive them before submitting an attempt."""

    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    question_text: str
    question_type: str
    options: dict | None = None
    order: int
    points: int


class QuizDetail(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    title: str
    passing_score: int
    questions: list[QuizQuestionPublic] = []


class QuizAttemptSubmit(BaseModel):
    answers: dict[str, str]  # question_id (as str) -> submitted answer


class QuizAttemptResult(BaseModel):
    score: float
    passed: bool
    correct_count: int
    total_questions: int

File: frontend/src/types/curriculum.ts
export interface LearningPathSummary {
  id: string
  title: string
  slug: string
  description: string | null
}

export interface CourseSummary {
  id: string
  title: string
  slug: string
  description: string | null
  level: string
  estimated_hours: number | null
}

export interface LearningPathDetail extends LearningPathSummary {
  courses: CourseSummary[]
}

export interface LessonSummary {
  id: string
  title: string
  slug: string
  order: number
  estimated_minutes: number | null
}

export interface ModuleWithLessons {
  id: string
  title: string
  slug: string
  order: number
  lessons: LessonSummary[]
}

export interface CourseDetail extends CourseSummary {
  modules: ModuleWithLessons[]
  completed_lesson_ids: string[]
}

export interface Skill {
  id: string
  name: string
  slug: string
  category: string | null
}

export type ContentBlock =
  | { type: 'objectives'; items: string[] }
  | { type: 'explanation'; beginner: string; technical: string }
  | { type: 'code'; language: string; code: string; output?: string }
  | { type: 'exercise'; prompt: string; starter_code: string }
  | { type: 'common_mistakes'; items: string[] }
  | { type: 'summary'; text: string }
  | { type: 'key_terms'; items: string[] }

export interface LessonDetail {
  id: string
  title: string
  slug: string
  order: number
  content: { blocks: ContentBlock[] }
  estimated_minutes: number | null
  skills: Skill[]
  quiz_id: string | null
}

export interface QuizQuestionPublic {
  id: string
  question_text: string
  question_type: string
  options: { choices?: string[] } | null
  order: number
  points: number
}

export interface QuizDetail {
  id: string
  title: string
  passing_score: number
  questions: QuizQuestionPublic[]
}

export interface QuizAttemptResult {
  score: number
  passed: boolean
  correct_count: number
  total_questions: number
}
